### PR TITLE
Fix float64 comparison test failure on archs using FMA

### PR DESCRIPTION
**Pull request by sbunce**
Background
=========
Architectures using FMA optimization yield slightly different results so we cannot assume floating point values will be precisely the same across different architectures.
https://github.com/prometheus/client_golang/pull/899#issuecomment-1244506390

Change
======
The solution in this change is to check "abs(a-b) < tolerance" instead of comparing the exact values. This will give us confidence that the histogram buckets are near identical.

Testing
=====
I have not tested this change on one of the previously failing architectures. Help from someone with access to one of these architectures would be appreciated.

### CODE REVIEW

File: prometheus/histogram_test.go
Comment (dswarbrick) [inline prometheus/histogram_test.go:362]: Wouldn't it make sense to avoid the repetition of the tolerance constant value here?
```suggestion
	if !equalFloat64s(got, want, tolerance) {
```

Comment (sbunce) [reply prometheus/histogram_test.go]: Oops, yes. That should be used.

**Comment by beorn7** [comment]
There are better ways of a tolerant comparison. We already have https://github.com/prometheus/prometheus/blob/main/promql/test.go#L637 in the Prometheus code base. Perhaps that's easy enough to just copy over here.

**Comment by beorn7** [comment]
Or depend on https://github.com/beorn7/floats :o)

**Comment by dswarbrick** [comment]
> Or depend on https://github.com/beorn7/floats :o)

I would prefer not to have to package yet another tiny dependency to make this build on Debian (and derivatives), so I would advocate just copying that function.

Or use something like https://pkg.go.dev/github.com/stretchr/testify/assert?utm_source=godoc#InDeltaSlice

**Comment by sbunce** [comment]
Sounds like there's consensus on avoiding the extra dependency. Reminds me of that go proverb, "a little copying is better than a little dependency".

I copied the contents of https://github.com/beorn7/floats to the prometheus/internal package. I made a few small changes.
- I renamed "ε" to "epsilon". I feel like this improves readability for non-math people slightly. I can switch back to the original variable name if people want.
- I modified the comment to reference the original repo.
- I added variants which operate on slices of floats.

Couple points I anticipate there may be disagreement about. Calling these out.
- I haven't copied over the tests. Seems like vendoring doesn't usually grab the tests, but this is a slightly different situation.
- I copied over the float32 variant of things, even though we don't use it. Let me know if I should remove those.

File: prometheus/internal/difflib.go
Comment (beorn7) [inline prometheus/internal/difflib.go:50]: I wouldn't put this in difflib.go because that file is really a (possibly modified) copy of https://github.com/pmezard/go-difflib/blob/master/difflib/difflib.go . Better to use a separate file for the AlmostEqual code.

**Review by beorn7 — COMMENTED** [review]
Looks good to me in general. Modulo location in file, see comment. Maintainers need to make the final call. @bwplotka @kakkoyun

**Comment by beorn7** [comment]
> I renamed "ε" to "epsilon".

I love the ε, but I am well aware that I'm a small minority here. (Insert lament about decline of classical education…)

**Comment by beorn7** [comment]
> I copied over the float32 variant of things.

I think it is overwhelmingly likely that we'll never use the float32 here. I would prefer to remove this variant. It's easy to add later should we really need it, thanks to the reference back to the source repo.

All other points you made seem fine to me.

Thanks for your contribution and your patience.

**Comment by sbunce** [comment]
You're welcome. I appreciate your feedback.

- I removed the float32 variants. I bet you're right that this will never be used.
- I moved the function to a new "almost_equal.go" file. I didn't realize the difflib file was a fork of a vendored dependency. Makes sense to not mix those.

**Review by beorn7 — APPROVED** [review]
Thanks again.

@bwplotka @kakkoyun please have a final look.

**Review by kakkoyun — APPROVED** [review]
LGTM :) Thanks for review @beorn7

**Pull request merged**
